### ISSUE TITLE
replace std.net.curl

### ISSUE THREAD

**Issue by wilzbach** [ikod/dlang-requests#10]
I am pretty sure you are aware of [this thread](http://forum.dlang.org/post/nhquzxmhwvawlxatzytr@forum.dlang.org). 

Many people want to get rid off the external dependency to curl, however 

> We need a replacement before we deprecate curl.

So let's join forces to create a great replacement?

@ikod if you are interested to go this road, you would need to relicense your library under the Boost license.

Ping at the usual people: @burner @JackStouffer @ZombineDev @schveiguy @CyberShadow @DmitryOlshansky @GilesBathgate - have you given this lib already a try?
A good way to help out is probably bug all design flaws and stuff you don't like as issues ;-)

**Comment by GilesBathgate** [comment]
@wilzbach This looks good: https://github.com/adamdruppe/arsd/blob/master/http.d any comments @adamdruppe

**Comment by ikod** [comment]
Hello,

I'm interested in writing good quality library first of all. If result will be considered as curl replacement, I'll be glad, of course.

I switched to Boost license.

**Comment by adamdruppe** [comment]
On Sun, May 15, 2016 at 04:02:28AM -0700, Giles Bathgate wrote:
> @wilzbach This looks good: https://github.com/adamdruppe/arsd/blob/master/http.d any comments @adamdruppe

I only have a minute right now but that one is obsoleted by http2:

https://github.com/adamdruppe/arsd/blob/master/http2.d


problem is I still haven't implemented SSL on Windows! I want to do
it without OpenSSL and use SChannel instead, so it just works out of
the box with the operating system library.

After that though, some cleanup and it might be useful. I'm using it
instead of curl on Linux now (and still finding bugs, but getting there)

**Comment by GilesBathgate** [comment]
@adamdruppe writing ssl/tls from scratch could be a challenge. A few years ago I experimented with polarssl which was a lightweight c library that does it. It was quite nice since most of the c files were self contained and could be used as separate modules. Well polarssl became ARM mbedtls and the source is available here: https://github.com/ARMmbed/mbedtls/tree/development/library porting that to D might be easier that starting from scratch. Unless of course D already has SSL/TLS and I missed something.

**Comment by bubnenkoff** [comment]
@adamdruppe <https://github.com/adamdruppe>  @Giles D already have native
Botan lob

2016-05-15 16:34 GMT+03:00 Giles Bathgate <notifications@github.com>:

**Comment by GilesBathgate** [comment]
@bubnenkoff Well I couldn't find it...I still can't. :(

**Comment by GilesBathgate** [comment]
@bubnenkoff Ah wait, its here: https://code.dlang.org/packages/botan So I guess that has to become part of phobos?

**Comment by bubnenkoff** [comment]
I hope...
For now Igor (ikod) work on support of it instead of openssl

воскресенье, 15 мая 2016 г. пользователь Giles Bathgate написал:

**Comment by JackStouffer** [comment]
The replacement should definitly take large amounts of "inspiration" from Python's [requests](http://docs.python-requests.org/en/master/). It's the most popular Python library for a reason.

**Comment by adamdruppe** [comment]
On Sun, May 15, 2016 at 06:34:16AM -0700, Giles Bathgate wrote:
> @adamdruppe writing ssl/tls from scratch could be a challenge.

Indeed, that's not my plan. I just want to use the native library
on Windows.

On Linux, using OpenSSL is basically the native lib, since that's
the interface that is always packaged with the operating system. But
on Windows, OpenSSL is a third-party dependency. A library called schannel
is the native solution there, included with the Win32 API.

(though there's also a full-blown HTTP implementation included with Win32,
so arguably we could just use it... but I want to solve the generic ssl
problem anyway.)

**Comment by ikod** [comment]
Hello,

Yes, I started to look at the botan, and hope I will be able to use it in
the nearest future.
But before I'd like to finish FTP "get/post" requests, as this can require
some code reorganisation. If this part wil be done right, then it will be
possible to plug in another protocols.

FTP code can take one or two weeks for completion.

2016-05-15 16:41 GMT+03:00 Dmitry Bubnenkov <notifications@github.com>:

**Comment by adamdruppe** [comment]
So, what my http2 does do right now:

* url building,
* automatic cookie handling
* connection pooling
* semi-async model (when you send the request, you get back a waitable object)

What is coming somewhat soon:

* web api consumer class (I did HipChat and Asana clients recently and found they were very similar so the library should abstract that even more)
* there's still a number of bugs

**Comment by GilesBathgate** [comment]
@adamdruppe Yes some of the connection pooling and cookie handling would be good to add to @ikod 's library. (I couldn't see anything like that already, correct me if I am wrong) Python requests has the session class to handle these matters, perhaps it could be encapsulated in a similar way.

**Comment by ikod** [comment]
@GilesBathgate, you are right, no pooling and cookie. Cookie are in the TODO.
My closest priorities are:
1. finish FTP
2. look close at the botan and try to use it (looks like vibe.d already use it)
3. implement cookies
4. allow user callbacks for get/post data streams (to support progress bar for example)
5. Provide high-level functions for file download (resume transfer, ETag and If-Modified-Since handling, etc)

Thanks for mentioning pooling, will keep it in mind.

**Comment by GilesBathgate** [comment]
> 1. finish FTP
@ikod @wilzbach Since you are proposing this to be a replacement for std.net.curl, don't forget you will need to also implement:

 DICT, FILE,  FTPS, Gopher, HTTPS, IMAP, IMAPS, LDAP, LDAPS, POP3, POP3S, RTMP, RTSP, SCP, SFTP, SMTP, SMTPS, Telnet and TFTP.

Assuming those are actually exposed by std.net.curl.

**Comment by bubnenkoff** [comment]
I do not think that all of them are really needed. Most of them should be in optional package.

What is FILE protocol? std.file ?

**Comment by GilesBathgate** [comment]
>What is FILE protocol? std.file ?
Yeah like ```file:///c:\file.txt```

Anyway I guess what I am getting at was already covered here: http://forum.dlang.org/post/mailman.1951.1462784155.26339.digitalmars-d@puremagic.com

**Comment by GilesBathgate** [comment]
This uri abstraction might be a useful sub component of these http libs: https://github.com/rikkimax/alphaPhobos/blob/master/source/std/experimental/uri.d

**Comment by ikod** [comment]
Hello,

Will check URI parser.
Re exotic protocols - I agree, curl in form of dub package will do the job
nicely.


2016-05-16 15:42 GMT+03:00 Giles Bathgate <notifications@github.com>:

**Comment by adamdruppe** [comment]
On Mon, May 16, 2016 at 05:42:03AM -0700, Giles Bathgate wrote:
> This uri abstraction might be a useful sub component of these http libs: https://github.com/rikkimax/alphaPhobos/blob/master/source/std/experimental/uri.d

I also have a URI component in my http2.d... that last check, is more standards-compliant and has more useful features than that one, though he's still working on it too.

**Comment by adamdruppe** [comment]
On Mon, May 16, 2016 at 03:55:31AM -0700, Giles Bathgate wrote:
>  DICT, FILE,  FTPS, Gopher, HTTPS, IMAP, IMAPS, LDAP, LDAPS, POP3, POP3S, RTMP, RTSP, SCP, SFTP, SMTP, SMTPS, Telnet and TFTP.

I know SMTP and SMTPS are exposed by std.net.curl.... it is actually
the only function in there that I actually use!

The others may be useful at times, but I don't  think the high level
library exposes them.

**Comment by wilzbach** [comment]
> Assuming those are actually exposed by std.net.curl.

Luckily only HTTP,  FTP and SMTP are exposed :)

http://dlang.org/phobos/std_net_curl.html

**Comment by GilesBathgate** [comment]
That's good, ```file:///``` would be nice, but better to get the minimum working first.

**Comment by ikod** [comment]
Hello,
basic cookie handling ready. What is missed - cookie persistency and expiration. Cookie lives as long as Request, but application have access to cookies - can retrieve them(and store) and set for new Request.

About the python-requests Session: what should be implemented?
By default, all dlang-requests are persistent and should correctly handle protocol/port/host/e.t.c switching. So this code 
```
Request rq = Request();
rq.get("http://httpbin.org/a");
rq.post("http://httpbin.org/b");
rq.get("ftp://ftp.example.com/pub/readme");
```
will use single connection to httpbin.org which will be reopened for FTP transfer.
What might be interesting - add some kind of parallelism on request to multiple url's. For example:
```
auto urls = [
    "http://httpbin.org/a",
    "http://httpbin.org/b",
    "ftp://example.com",
    "http://httpbin.org/c"
];
Response[] rs = rq.muliGet(urls);
```
Optimizing this can require sorting urls and running them in parallel to maximize connection reuse.

**Comment by GilesBathgate** [comment]
As I understand, the python session API is useful for maintain state across http calls, via cookies, since HTTP is stateless. It would probably be useful to provide the ability to maintain state this way in addition to being able to re-use connections. Fundamentally though the API should expose a stateless API as this is probably the most common use case.

**Comment by ikod** [comment]
Yes, this how it works now in dlang requests.
In the code:
```
Request rq = Request();
rq.get("http://httpbin.org/a");
rq.get("ftp://ftp.example.com/pub/readme");
rq.post("http://httpbin.org/b");
```
if server send Cookie-Set on the first call (to "http://httpbin.org/a"), then cookies for domain httpbin.org will be used for each call as long as *rq* instance used.

**Comment by ikod** [comment]
Hello,

This is just current state report.

I implemented most of the functionality I planned from start, and currently I'm fixing minor bugs and add minor features.
What is missing and broken:
1. unit-testing rely on internet access to httpbin.org. Fixing this and enable testis using internal http server is my next goal.
2. docs are not complete and buggy both in content and style.

Your notes are more than welcome,

Thanks!

**Comment by ikod** [comment]
Hello @GilesBathgate 

just rediscovered this old issue.

> As I understand, the python session API is useful for maintain state across http calls, via cookies, since HTTP is stateless. It would probably be useful to provide the ability to maintain state this way in addition to being able to re-use connections. Fundamentally though the API should expose a stateless API as this is probably the most common use case.

It was implemented recently - connection pooling/reusing (also 'interceptors'). Now `Request` struct looks very much like Session for python requests. It encapsulate all configuration parameters for connections, as long all persistent things (connections, permanent redirects, cookies, ...)

I'll probably close this issue.
